l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,720 @@
+    /// @param _prevMsgSender The address of the entity that initiated the deposit.
+    /// @param _assetId The unique identifier of the deposited L1 token.
+    /// @param _transferData The encoded transfer data, which includes both the deposit amount and the address of the L2 receiver.
+    /// @param _expectedTxDataHash The nullifier data hash stored for the failed deposit.
+    /// @return isLegacyTxDataHash True if the transaction is legacy, false otherwise.
+    function _isLegacyTxDataHash(
+        address _prevMsgSender,
+        bytes32 _assetId,
+        bytes memory _transferData,
+        bytes32 _expectedTxDataHash
+    ) internal view returns (bool isLegacyTxDataHash) {
+        try this.encodeTxDataHash(LEGACY_ENCODING_VERSION, _prevMsgSender, _assetId, _transferData) returns (
+            bytes32 txDataHash
+        ) {
+            return txDataHash == _expectedTxDataHash;
+        } catch {
+            return false;
+        }
+    }
+
+    /// @dev Determines if a deposit was initiated on ZKsync Era before the upgrade to the Shared Bridge.
+    /// @param _chainId The chain ID of the transaction to check.
+    /// @param _l2BatchNumber The L2 batch number for the deposit where it was processed.
+    /// @param _l2TxNumberInBatch The L2 transaction number in the batch, in which the deposit was processed.
+    /// @return Whether deposit was initiated on ZKsync Era before Shared Bridge upgrade.
+    function _isEraLegacyDeposit(
+        uint256 _chainId,
+        uint256 _l2BatchNumber,
+        uint256 _l2TxNumberInBatch
+    ) internal view returns (bool) {
+        if ((_chainId == ERA_CHAIN_ID) && (eraLegacyBridgeLastDepositBatch == 0)) {
+            revert SharedBridgeValueNotSet(SharedBridgeKey.LegacyBridgeLastDepositBatch);
+        }
+        return
+            (_chainId == ERA_CHAIN_ID) &&
+            (_l2BatchNumber < eraLegacyBridgeLastDepositBatch ||
+                (_l2TxNumberInBatch <= eraLegacyBridgeLastDepositTxNumber &&
+                    _l2BatchNumber == eraLegacyBridgeLastDepositBatch));
+    }
+
+    /// @notice Verifies the validity of a withdrawal message from L2 and returns withdrawal details.
+    /// @param _finalizeWithdrawalParams The structure that holds all necessary data to finalize withdrawal
+    /// @return assetId The ID of the bridged asset.
+    /// @return transferData The transfer data used to finalize withdawal.
+    function _verifyWithdrawal(
+        FinalizeL1DepositParams calldata _finalizeWithdrawalParams
+    ) internal view returns (bytes32 assetId, bytes memory transferData) {
+        (assetId, transferData) = _parseL2WithdrawalMessage(
+            _finalizeWithdrawalParams.chainId,
+            _finalizeWithdrawalParams.message
+        );
+        L2Message memory l2ToL1Message;
+        {
+            bool baseTokenWithdrawal = (assetId == BRIDGE_HUB.baseTokenAssetId(_finalizeWithdrawalParams.chainId));
+            if (baseTokenWithdrawal) {
```

Comment:
I don't understand why I can't forge the deposit by simply sending message in a required format from any L2 address?